Improve tests a little (#1324)
gbdev/rgbds: CONTRIBUTING.md

6. Create a pull request against the branch `master`.
7. Be prepared to get some comments about your code and to modify it. Tip: Use
   `git rebase -i origin/master` to modify chains of commits.

## Adding a test

The test suite is a little ad-hoc, so the way tests work is different for each program being tested.

Feel free to modify how the test scripts work, 

### RGBASM

Each `.asm` file corresponds to one test.

> RGBASM will be invoked on the `.asm` file.

Reviewer: ```suggestion
RGBASM will be invoked on the `.asm` file with all warnings enabled.
```

> If a `.out` file exists, RGBASM's output (`println` etc.) must match its contents.

Reviewer: ```suggestion
If a `.out` file exists, RGBASM's output (`print`, `println`, etc.) must match its contents.
```

> If a `.err` file exists, RGBASM's errors (`warn`, errors, etc.) must match its contents.

Reviewer: (Clarifying "errors" by listing "errors" sounded odd.)
```suggestion
If a `.err` file exists, RGBASM's error output (`warn`, errors, etc.) must match its contents.
```

If a `.out.bin` file exists, the object file will be linked, and the generated ROM truncated to the length of the `.out.bin` file.
After that, the ROM must match the `.out.bin` file.

> 

Reviewer: Please also explain that you can generate your output files something like this:
```bash
# Enter the test/asm directory
pushd tests/asm
# Generate `.out` and `.err` files
../../rgbasm -Weverything -o tmp.o mytest.asm >mytest.out 2>mytest.err
# Generate `.out.bin` file
../../rgblink -x -o mytest.out.bin tmp.o
# Delete mytest.out and/or mytest.err if they have no content
! grep -q . mytest.out && rm mytest.out
! grep -q . mytest.err && rm mytest.err
# Cleanup
rm tmp.o
popd
```

Author: I'd rather direct people at creating empty files, checking if the output makes sense, and copy-pasting it if so. I think it's better, because it provides some time to pause and think about that.

### RGBLINK

Each `.asm` file corresponds to one test, or one *set* of tests.

All tests begin by assembling the `.asm` file into an object file, which will be linked in various ways depending on the test.

> 

Reviewer: Likewise for how to make the expected output of RGBLINK tests (they do have `.out` and `.out.bin` files).

Author: See https://github.com/gbdev/rgbds/pull/1324/files#r1509967319

#### Simple tests

> These simply check that RGBLINK's output matches... *something*.

Reviewer: ```suggestion
These simply check that RGBLINK's output matches some expected output.

```

A `.out` file **must** exist, and RGBLINK's output must match that file's contents.

Additionally, if a `.out.bin` file exists, the `.gb` file generated by RGBLINK must match it.

#### Linker script tests

These allow applying various linker scripts to the same object file.
If one or more `.link` files exist, whose names start the same as the `.asm` file, then each of those files correspond to one test.

Each `.link` linker script **must** be accompanied by a `.out` file, and RGBLINK's output must match that file's contents when passed the corresponding linker script.

#### Variant tests

These allow testing RGBLINK's `-d`, `-t`, and `-w` flags.

> If one or more files called `<asm file name>-<flag>.out` or `<asm file name>-no-<flag>.out` exist, then each of them corresponds to one test.

Reviewer: In general:
```
`foo-<var>-bar`
```
should be:
```
<code>foo-<i>&lt;var&gt;</i>-bar</code>
```
(RELEASE.md has some cases.)

Author: Fixed by 48479fee, though not in the two code *blocks* at the end for the reason explained in the commit message.

> If one or more files called <code><var>&lt;asm file name&gt;</var>-<var>&lt;flag&gt;</var>.out</code> or <code><var>&lt;asm file name&gt;</var>-no-<var>&lt;flag&gt;</var>.out</code> exist, then each of them corresponds to one test.

Reviewer: Since the rest of the doc says e.g. `.out.bin` and not <code><var>&lt;asm file name&gt;</var>.out.bin</code>.
```suggestion
If one or more <code>-<var>&lt;flag&gt;</var>.out</code> or <code>-no-<var>&lt;flag&gt;</var>.out</code> files exist, then each of them corresponds to one test.
```

The object file will be linked with and without said flag, respectively; and in each case, RGBLINK's output must match the `.out` file's contents.

### RGBFIX

Each `.bin` file corresponds to one test, and **must** be accompanied by a `.flags` file and a `.err` file.

The `.flags` file is a text file whose first line contains flags to pass to RGBFIX.

> (There may be more lines, which will be ignored; they can serve as commentsto explain what the test is about.)

Reviewer: ```suggestion
(There may be more lines, which will be ignored; they can serve as comments to explain what the test is about.)
```

RGBFIX will be invoked on the `.bin` file, and its error output must match the contents of the `.err` file.
(If no errors ought to be printed, then the `.err` file should just be empty.)

Additionally, if a `.gb` file exists, the output of RGBFIX must match the `.gb`.

### RGBGFX

Each `.png` file corresponds to one test.
RGBGFX will be invoked on the file.
If a `.flags` file exists, it will be used as part of the RGBGFX invocation (<code>@<var>&lt;file&gt;</var>.flags</code>).

If no `.err` file exists, RGBGFX is simply expected to be able to process the file normally.
If one *does* exist, RGBGFX's return status is ignored, but its output **must** match the `.err` file's contents.

### Downstream projects

> 1. Make sure the downstream project supports <code>make <var>&lt;target&gt;</var> RGBDS=<var>&lt;path to RGBDS&gt;</var></code>.
>    While the test suite supports any Make target name, only [Make](//gnu.org/software/make) is currently supported, and the Makefile must support a `RGBDS` variable to use a non-system RGBDS.

Reviewer: ```suggestion
1. Make sure the downstream project supports <code>make <var>&lt;target&gt;</var> RGBDS=<var>&lt;path/to/RGBDS/&gt;</var></code>.
   While the test suite supports any Make target name, only [Make](//gnu.org/software/make) is currently supported, and the Makefile must support a `RGBDS` variable to use a non-system RGBDS directory.
```

>    Only projects hosted on GitHub are currently supported, too.

Reviewer: ```suggestion
   Also, only projects hosted on GitHub are currently supported.
```

2. Add the project to `test/fetch-test-deps.sh`: add a new `action` line at the bottom, following the existing pattern:
   
   ```sh

>    action <owner>/<repo>   <date of last commit>   <hash of last commit>

Reviewer: ```suggestion
   action  <owner>/<repo>  <date of last commit>  <hash of last commit>
```

   ```

   (The date is used to avoid fetching too much history when cloning the repositories.)
3. Add the project to `test/run-tests.sh`: add a new `test_downstream` line at the bottom, following the existing pattern:

   ```sh

>    test_downstream <repo>  <Make target>

Reviewer: ```suggestion
   test_downstream  <repo>  <make target>
```

Author: Perhaps
```suggestion
   test_downstream <repo>  <makefile target>
```
then?

Reviewer: I thought it was "`make` target" as in "target of the `make` command", but sure.

Author: Correct, but without the extra formatting, I don't think it's clear enough that this is referring to the command and not the English verb. (Hence my original capitalisation, to try drawing attention to that.)

Since the targets are defined in the Makefile, I think both formulations are acceptable, but "makefile target" is less ambiguous.

Reviewer: Whichever you prefer, then.

>    test_downstream <repo>  <makefile target>

Reviewer: ```suggestion
   test_downstream  <repo>  <makefile target>
```
